Импорт данных для обучения и конфигурация модели

In [41]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

%matplotlib inline
%config InlineBackend.figure_format = 'svg'

plt.style.use('fivethirtyeight')
warnings.filterwarnings('ignore')

In [42]:
data = pd.read_csv('train.csv', index_col = 0, encoding = 'utf-8')

Убедимся в том, что импорт данных прошел успешно

In [43]:
data.head()

DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
Id                                                                      
14038          35    2.0  47.981561   29.442751            6.0      7   
15053          41    3.0  65.683640   40.049543            8.0      7   
4765           53    2.0  44.947953   29.197612            0.0      8   
5809           58    2.0  53.352981   52.731512            9.0      8   
10783          99    1.0  39.649192   23.776169            7.0     11   

       HouseFloor  HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  \
Id                                                                      
14038         9.0       1969   0.089040         B         B        33   
15053         9.0       1978   0.000070         B         B        46   
4765         12.0       1968   0.049637         B         B        34   
5809         17.0       1977   0.437885         B         B        23   
10783        12.0       1976   0.012339         B         B        35   

       Social_2  Social_3  Healthcare_1  Helthcare_2  Shops_1 Shops_2  \
Id                                                                      
14038      7976         5           NaN            0       11       B   
15053     10309         1         240.0            1       16       B   
4765       7759         0         229.0            1        3       B   
5809       5735         3        1084.0            0        5       B   
10783      5776         1        2078.0            2        4       B   

               Price  
Id                    
14038  184966.930730  
15053  300009.450063  
4765   220925.908524  
5809   175616.227217  
10783  150226.531644

Проверим целостность данных

In [44]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 14038 to 6306
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DistrictId     10000 non-null  int64  
 1   Rooms          10000 non-null  float64
 2   Square         10000 non-null  float64
 3   LifeSquare     7887 non-null   float64
 4   KitchenSquare  10000 non-null  float64
 5   Floor          10000 non-null  int64  
 6   HouseFloor     10000 non-null  float64
 7   HouseYear      10000 non-null  int64  
 8   Ecology_1      10000 non-null  float64
 9   Ecology_2      10000 non-null  object 
 10  Ecology_3      10000 non-null  object 
 11  Social_1       10000 non-null  int64  
 12  Social_2       10000 non-null  int64  
 13  Social_3       10000 non-null  int64  
 14  Healthcare_1   5202 non-null   float64
 15  Helthcare_2    10000 non-null  int64  
 16  Shops_1        10000 non-null  int64  
 17  Shops_2        10000 non-null  object 
 18  Pri

In [45]:
from sklearn.model_selection import train_test_split, cross_val_score, RandomizedSearchCV
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score, mean_squared_error

Проведем минимальную чистку и предобработку данных
Построим несколько моделей и оценим качество на отложенной выборке по метрике R^2
Из построенных выберем лучшую baseline модель

In [46]:
cleared_data = data.fillna(0)

In [47]:
X = cleared_data.drop(labels = ['Price'], axis = 1)
X = pd.get_dummies(data = X, columns = ['DistrictId', 'Ecology_2', 'Ecology_3', 'Shops_2'], drop_first= True)
y = cleared_data['Price'].values

In [48]:
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size = 0.3, random_state = 42, shuffle = True)

In [49]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_valid_scaled = scaler.transform(X_valid)

'lasso': Lasso(random_state = 42),
'ridge': Ridge(random_state= 42),
'enet': ElasticNet(random_state=42).

In [50]:
model_dict = {
   'lr': LinearRegression(),
   'knn': KNeighborsRegressor(),
   'dt': DecisionTreeRegressor(random_state = 42),
   'rf': RandomForestRegressor(random_state = 42)
}

score_list = []

for model in model_dict:
    model_dict[model].fit(X_train_scaled, y_train)
    y_pred_train = model_dict[model].predict(X_train_scaled)
    y_pred_valid = model_dict[model].predict(X_valid_scaled)
    score_train = r2_score(y_train, y_pred_train)
    score_valid = r2_score(y_valid, y_pred_valid)
    score_list.append((score_train, score_valid))

for model, scores in zip(model_dict.keys(), score_list):
    print(f'{model}: \n\ttrain: {np.round(scores[0], 4)} \n\tvalid: {np.round(scores[1], 4)}')

lr: 
	train: 0.6759 
	valid: 0.5829
knn: 
	train: 0.7495 
	valid: 0.5906
dt: 
	train: 1.0 
	valid: 0.5294
rf: 
	train: 0.9612 
	valid: 0.716


Лучший результат после первых доработок оказался RandomForest.
BaseLine R^2 = 0.716
Далее чистка данных и продолжение работы с RandonForest.

In [51]:
cleared_data = data[data['Rooms'] < 10]
cleared_data = cleared_data[cleared_data['Square'] < 250]
cleared_data = cleared_data[cleared_data['HouseFloor'] < 100]

Функция предобработки.

In [52]:
def preproces(df_input, dummy_features = None):
    
    df_output = df_input.copy()
        
    Square = dict(df_output.groupby('Rooms')['Square'].median())
    idx = (df_output['Square'] < 15) & (df_output['Rooms'] < 2)
    df_output.loc[idx, 'Square'] = df_output.loc[idx, 'Rooms'].apply(lambda x: Square[x])
    
    idx = (df_output['Square'] < 30) & (df_output['Rooms'] > 1)
    df_output.loc[idx, 'Square'] = df_output.loc[idx, 'Rooms'].apply(lambda x: Square[x])
    
    idx = (df_output['Square'] < df_output['LifeSquare'])
    df_output.loc[idx, 'Square'] = df_output.loc[idx, 'LifeSquare']

    LifeSquare = dict(df_output.groupby('Rooms')['LifeSquare'].median())
    idx = (df_output['LifeSquare'].isnull())
    df_output.loc[idx, 'LifeSquare'] = df_output.loc[idx, 'Rooms'].apply(lambda x: LifeSquare[x])

    idx = (df_output['Rooms'] < 1)
    df_output.loc[idx, 'KitchenSquare'] = 0.0
    
    KitchenSquare = dict(df_output.groupby('Rooms')['KitchenSquare'].median())
    idx = (df_output['Rooms'] > 0)
    df_output.loc[idx, 'KitchenSquare'] = df_output.loc[idx, 'Rooms'].apply(lambda x: KitchenSquare[x])
    
    idx = (df_output['HouseFloor'] < df_output['Floor'])
    df_output.loc[idx, 'HouseFloor'] = df_output.loc[idx, 'Floor']
    
    idx = (df_output['HouseYear'] >= 2019)
    df_output.loc[idx, 'HouseYear'] = 2019
   
    df_output['Healthcare_1'].fillna(df_output['Healthcare_1'].median(), inplace=True)
    
    df_output['Ecology_2'] = df_output['Ecology_2'].apply(lambda x: 1 if x == 'A' else 0)
    df_output['Ecology_3'] = df_output['Ecology_3'].apply(lambda x: 1 if x == 'A' else 0)
    df_output['Shops_2'] = df_output['Shops_2'].apply(lambda x: 1 if x == 'A' else 0)
    
    if dummy_features != None:
        df_output = pd.get_dummies(data = df_output, columns = dummy_features, drop_first = True)
    
    return df_output

Предобработаем данные.

In [53]:
processed_data = preproces(cleared_data, dummy_features = None)

In [54]:
X = processed_data.drop(labels=['DistrictId', 'Price'], axis=1)
y = processed_data['Price'].values

In [55]:
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size = 0.3, random_state = 42, shuffle = True)

In [56]:
scaler = MinMaxScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_valid_scaled = scaler.transform(X_valid)

Применим настройку с помощью RandomizedSearchCV

In [57]:
rf = RandomForestRegressor(random_state = 42, n_jobs=-1)

random_grid = {
    'n_estimators': np.arange(200, 501, 20),
    'max_depth': np.arange(2, 51, 2),
    'max_features': [0.5, 0.6, 0.7, 0.8, 0.9],
    'min_samples_leaf': [1, 2, 4],
    'min_samples_split': [2, 5, 10]
}

random_search = RandomizedSearchCV(
    estimator = rf,
    param_distributions = random_grid,
    n_iter = 50,
    scoring = 'r2',
    cv = 10,
    verbose = 2,
    random_state = 42,
    n_jobs=-1
)

random_search.fit(X_train_scaled, y_train)

print(random_search.best_score_)
print(random_search.best_params_)

Fitting 10 folds for each of 50 candidates, totalling 500 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  25 tasks      | elapsed:   22.2s
[Parallel(n_jobs=-1)]: Done 146 tasks      | elapsed:  1.8min
[Parallel(n_jobs=-1)]: Done 349 tasks      | elapsed:  4.4min
[Parallel(n_jobs=-1)]: Done 500 out of 500 | elapsed:  6.3min finished


0.7248310368651174
{'n_estimators': 400, 'min_samples_split': 5, 'min_samples_leaf': 1, 'max_features': 0.5, 'max_depth': 26}


In [58]:
y_pred_train = random_search.best_estimator_.predict(X_train_scaled)
y_pred_valid = random_search.best_estimator_.predict(X_valid_scaled)

score_train = r2_score(y_train, y_pred_train)
score_valid = r2_score(y_valid, y_pred_valid)

print(f'train: {np.round(score_train, 4)}\nvalid: {np.round(score_valid, 4)}')

train: 0.9377
valid: 0.7165


Обучение

In [63]:
scaler = MinMaxScaler()

X_scaled = scaler.fit_transform(X)

model = RandomForestRegressor(
    n_estimators = 400,
    max_depth = 26,
    max_features = 0.5,
    min_samples_leaf = 1,
    min_samples_split = 5,
    random_state = 42,
    n_jobs=-1
)

model.fit(X_scaled, y)

scores = cross_val_score(model, X_scaled, y, scoring = 'r2', cv = 10, n_jobs=-1)

print(f'cv mean: {np.round(np.mean(scores), 4)}')
print(f'cv std:  {np.round(np.std(scores), 4)}')

cv mean: 0.7351
cv std:  0.0243


Новые (неразмеченные) данные

In [64]:
new_data = pd.read_csv('test.csv', index_col = 0, encoding = 'utf-8')

In [65]:
new_data.head()

DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
Id                                                                      
725            58    2.0  49.882643   33.432782            6.0      6   
15856          74    2.0  69.263183         NaN            1.0      6   
5480          190    1.0  13.597819   15.948246           12.0      2   
15664          47    2.0  73.046609   51.940842            9.0     22   
14275          27    1.0  47.527111   43.387569            1.0     17   

       HouseFloor  HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  \
Id                                                                      
725          14.0       1972   0.310199         B         B        11   
15856         1.0       1977   0.075779         B         B         6   
5480          5.0       1909   0.000000         B         B        30   
15664        22.0       2007   0.101872         B         B        23   
14275        17.0       2017   0.072158         B         B         2   

       Social_2  Social_3  Healthcare_1  Helthcare_2  Shops_1 Shops_2  
Id                                                                     
725        2748         1           NaN            0        0       B  
15856      1437         3           NaN            0        2       B  
5480       7538        87        4702.0            5        5       B  
15664      4583         3           NaN            3        3       B  
14275       629         1           NaN            0        0       A

Обработка новых данных

In [67]:
processed_new_data = preproces(new_data, dummy_features = None)

In [68]:
X_new = processed_new_data.drop(labels = ['DistrictId'], axis = 1)

Масштабируем данные при помощи MinMaxScaler

In [70]:
X_new_scaled = scaler.transform(X_new)

Процесс предскозания

In [71]:
y_pred_new = model.predict(X_new_scaled)

Запишем информацию в DataFrame

In [72]:
predictions = pd.DataFrame({
    'Id': X_new.index,
    'Price': y_pred_new
})

In [73]:
predictions.head()

Id          Price
0    725  161277.175142
1  15856  203216.976711
2   5480  276090.829338
3  15664  351466.498501
4  14275  136403.648400

In [74]:
predictions.to_csv('Smirnykh_predictions.csv', sep=',', index = False, encoding = 'utf-8')